pkg/extensions/tbtc/derive_address.go

```
@@ -0,0 +1,105 @@
+//
+// The returned address will be a p2pkh/p2sh address for prefixes xpub and tpub,
+// (i.e. prefixed by 1, m, or n), a p2wpkh-in-p2sh address for prefixes ypub or
+// upub (i.e., prefixed by 3 or 2), and a bech32 p2wpkh address for prefixes
+// zpub or vpub (i.e., prefixed by bc1 or tb1).
+//
+// See [BIP32], [BIP44], [BIP49], and [BIP84] for more on address derivation,
+// particular paths, etc.
+//
+// [BIP32]: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
+// [BIP44]: https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki
+// [BIP49]: https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki
+// [BIP84]: https://github.com/bitcoin/bips/blob/master/bip-0084.mediawiki
+func DeriveAddress(extendedPublicKey string, addressIndex int) (string, error) {
+	extendedKey, err := hdkeychain.NewKeyFromString(extendedPublicKey)
+	if err != nil {
+		return "", fmt.Errorf(
+			"error parsing extended public key: [%s]",
+			err,
+		)
+	}
+	// For later usage---this is xpub/ypub/zpub/...
+	publicKeyDescriptor := extendedPublicKey[0:4]
+
+	externalChain := extendedKey
+	for externalChain.Depth() < 4 {
+		// Descend the hierarchy at /0 until the external chain path, `m/*/*/*/0`.
+		// ex: If we get a `m/32'/5` extended key, we descend to `m/32'/5/0/0`.
+		externalChain, err = externalChain.Child(0)
+		if err != nil {
+			return "", fmt.Errorf(
+				"error deriving external chain path /0 from extended key: [%s]",
+				err,
+			)
+		}
+	}
+
+	requestedPublicKey, err := externalChain.Child(uint32(addressIndex))
+	if err != nil {
+		return "", fmt.Errorf(
+			"error deriving requested address index /0/%v from extended key: [%s]",
+			addressIndex,
+			err,
+		)
+	}
+
+	// Now to decide how we want to serialize the address...
+	var chainParams *chaincfg.Params
+	switch publicKeyDescriptor {
+	case "xpub", "ypub", "zpub":
+		chainParams = &chaincfg.MainNetParams
+	case "tpub", "upub", "vpub":
+		chainParams = &chaincfg.TestNet3Params
+	}
```

Should we throw an error on a not supported `publicKeyDescriptor` value, e.g. `yprv`, `Ypub`, etc.?